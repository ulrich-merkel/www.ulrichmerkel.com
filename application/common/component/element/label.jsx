/* eslint-disable immutable/no-mutation */
/**
 * Rendering a label html tag.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.2
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 *
 * @changelog
 * - 0.0.2 Add isVisuallyHidden to props
 * - 0.0.1 Basic functions and structure
 */
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

/**
 * Function representing a component to return a single react child element.
 *
 * @param {object} [props] - The current component props
 * @param {string} [props.htmlFor] - The label for attribute
 * @param {string} [props.className] - The label css class names, will be merged into component default classNames
 * @param {boolean} [props.isVisuallyHidden] - Whether the label is visually hidden or not
 * @returns {ReactElement} React component markup
 */
function ElementLabel(props) {
    const {
        className,
        htmlFor,
        isVisuallyHidden,
        ...otherProps
    } = props;

    const composedClassName = classnames(
        'm-form__label',
        className,
        {
            'is-visually-hidden': isVisuallyHidden
        }
    );

    return (
        // eslint-disable-next-line react/jsx-props-no-spreading, jsx-a11y/label-has-for,  jsx-a11y/label-has-associated-control
        <label className={composedClassName} htmlFor={htmlFor} {...otherProps} />
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 */
ElementLabel.propTypes = {
    htmlFor: PropTypes.string.isRequired,
    className: PropTypes.string, // eslint-disable-line react/require-default-props
    isVisuallyHidden: PropTypes.bool
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 */
ElementLabel.defaultProps = {
    isVisuallyHidden: false
};

export default ElementLabel;
