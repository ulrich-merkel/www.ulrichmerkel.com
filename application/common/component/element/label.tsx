/**
 * Rendering a label html tag.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, ReactNode } from 'react';
import classnames from 'classnames';

type Props = {
    children?: ReactNode;
    className?: string;
    htmlFor: string;
    isVisuallyHidden?: boolean;
};

/**
 * Function representing a html form label element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @param {string} [props.htmlFor] - The label for attribute
 * @param {string} [props.className] - The label css class names, will be merged into component default classNames
 * @param {boolean} [props.isVisuallyHidden] - Whether the label is visually hidden or not
 * @returns {ReactElement} React component markup
 */
export const Label: FunctionComponent<Props> = (props) => {
    const {
        className,
        htmlFor,
        isVisuallyHidden = false,
        ...otherProps
    } = props;

    const composedClassName = classnames('m-form__label', className, {
        'is-visually-hidden': isVisuallyHidden
    });

    return (
        // eslint-disable-next-line react/jsx-props-no-spreading, jsx-a11y/label-has-for,  jsx-a11y/label-has-associated-control
        <label
            className={composedClassName}
            htmlFor={htmlFor}
            {...otherProps}
        />
    );
};
