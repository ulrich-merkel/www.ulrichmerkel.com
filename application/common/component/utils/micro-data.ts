/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
