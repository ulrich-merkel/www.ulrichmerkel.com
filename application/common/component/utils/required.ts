/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
