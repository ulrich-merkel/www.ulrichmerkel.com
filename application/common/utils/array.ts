/**
 * Es6 module for handling array related utilities.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */

/**
 * Check if given "input" is a valid array with at least one entry.
 *
 * @param {Array<*>} array - The array to be checked
 * @returns {boolean} Whether this is an array with some entries
 */
export function isValidArray(array?: any[]): boolean {
    return Array.isArray(array) && array.length > 0;
}
