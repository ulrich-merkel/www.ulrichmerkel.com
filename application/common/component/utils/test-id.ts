/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
