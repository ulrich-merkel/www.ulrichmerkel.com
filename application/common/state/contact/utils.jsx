/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 *
 * @TODO Should be refactored and moved to actions
 *
 * @requires common/utils/validation
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';

/**
 * Validation helper to check all inputs.
 *
 * @function
 * @param {object} [state={}] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
function validate(state = {}) {
    return {
        name: required(state.name),
        email: required(state.email) && email(state.email),
        subject: required(state.subject),
        message: required(state.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @function
 * @param {object} state - The current redux contact state
 * @returns {boolean} The validation status
 */
function isValid(state) {
    const validated = validate(state);

    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

function canSendForm(state) {
    const isStateValid = isValid(state);

    return isStateValid
        && state.pristine
        && !state.pending
        && isBrowser(state);
}

export {
    canSendForm,
    isValid,
    validate
};
