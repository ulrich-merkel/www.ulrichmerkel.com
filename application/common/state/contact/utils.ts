/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';
import { ContactFormType, ValidateType } from './types';

/**
 * Validation helper to check all inputs.
 *
 * @param {object} [contactForm] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
export function validate(contactForm: ContactFormType): ValidateType {
    return {
        name: required(contactForm?.name),
        email: required(contactForm?.email) && email(contactForm?.email),
        subject: required(contactForm?.subject),
        message: required(contactForm?.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} The validation status
 */
export function isValid(contactForm: ContactFormType): boolean {
    const validated = validate(contactForm);
console.log({validated, contactForm})
    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

/**
 * Check if the form can be send.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} Whether the submit button is enabled or not
 */
export function canSendForm(contactForm: ContactFormType): boolean {
    const isStateValid = isValid(contactForm);

    return (
        isStateValid &&
        contactForm.pristine &&
        !contactForm.pending &&
        isBrowser(contactForm)
    );
}
