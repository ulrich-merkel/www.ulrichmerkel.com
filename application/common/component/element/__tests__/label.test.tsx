import * as React from 'react';
import { render } from '../../../../__tests__/utils/test-utils';
import { Label } from '../label';

describe('Label', function fnDescribe() {
    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <Label className="label" htmlFor="foo-bar">
                Label Children
            </Label>
        );
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render correctly the hidden state', function fnIt() {
        const { asFragment } = render(
            <Label className="label" htmlFor="foo-bar" isVisuallyHidden>
                Label Children
            </Label>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
