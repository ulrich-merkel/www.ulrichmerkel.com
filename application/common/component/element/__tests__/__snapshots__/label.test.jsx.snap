// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label should render correctly 1`] = `
<DocumentFragment>
  <label
    class="m-form__label label"
    for="foo-bar"
  >
    Label Children
  </label>
</DocumentFragment>
`;

exports[`Label should render correctly the hidden state 1`] = `
<DocumentFragment>
  <label
    class="m-form__label label is-visually-hidden"
    for="foo-bar"
  >
    Label Children
  </label>
</DocumentFragment>
`;
